app/controllers/oaisys/pmh_controller.rb

```diff
@@ -2,25 +2,84 @@
 
 module Oaisys
   class PMHController < ApplicationController
+    EXPECTED_ARGUMENTS = {
+        identify: {
+            required: %w(),
+            optional: %w(),
+            exclusive: %w()
+        },
+        list_sets: {
+            required: %w(),
+            optional: %w(),
+            exclusive: %w(resumptionToken)
+        },
+        list_metadata_formats: {
+            required: %w(),
+            optional: %w(identifier),
+            exclusive: %w()
+        },
+        list_records: {
+            required: %w(metadataPrefix),
+            optional: %w(from until set),
+            exclusive: %w(resumptionToken)
+        },
+        get_record: {
+            required: %w(identifier metadataPrefix),
+            optional: %w(),
+            exclusive: %w()
+        },
+        list_identifiers: {
+            required: %w(metadataPrefix),
+            optional: %w(from until set),
+            exclusive: %w(resumptionToken)
+        },
+    }.freeze
     def bad_verb
     end
 
     def identify
+      expected_ags(__method__)
     end
 
     def list_sets
+      expected_ags(__method__)
     end
 
     def list_metadata_formats
+      expected_ags(__method__)
     end
 
     def list_records
+      expected_ags(__method__)
     end
 
     def get_record
+      expected_ags(__method__)
     end
 
     def list_identifiers
+      expected_ags(__method__)
+    end
+
+    private
+
+    def expected_ags (verb)
```

Comment:
Looks like a typo? Should be `expected_args`.

I'm wondering if we shouldn't take this a bit further and call it `expect_args`, maybe with a signature like `expect_args(required:, optional:)` instead, because then we can make it a bit more DSL-y, eg)

```ruby
def list_identifiers
   expect_args required: :metadataPrefix, optional: [:from, :until, :set]

  # ....more logic here
end
```

That might be a little less magical, or at least require less scrolling back and forth to the hash, for future readers (aka us in a week) to figure out what's going on when we read an action. Thoughts?

Reply:
Done, the only downside I see to this approach is if we need to use the required, optional, and exclusive arguments elsewhere but that might not even happen

Reply:
We can definitely revisit this approach if we run into that. In general I think YAGNI is the best approach here -- if you don't know you need something and only think it's possible that someday you might, assume You Ain't Gonna Need It, and revisit if it does crop up.

(the alternate approach is to optimistically build a lot of abstractions ahead of time on the idea you might need them down the road, which seems nice if you do manage to correctly anticipate things, but in a lot of cases where you don't quite anticipate things right you just end up with a complicated abstraction that still needs refactoring anyways)